lib/std/json/static.zig

```diff
@@ -401,21 +401,25 @@ pub fn innerParse(
                                 i += slice.len;
                             },
                             .partial_string_escaped_1 => |arr| {
+                                if (arr.len > r.len) unreachable;
```

Comment:
if you bump this `if` down below the runtime length check, i think it would be good.

as is, these `unreachable`s can be hit if you parse `"\u2022"` into a `[1]u8`, for example (line 410). In general, if the utf8 encoding of an escape sequence is longer than then length of a fixed-size `[N]u8` array that is meant to store the string content, it will hit these unreachables.

this does look like a compiletime analysis limitation to me. if `i + arr.len > r.len` diverts control flow away from the `r[i..][0..arr.len]`, then the slicing is never going to be out of bounds; that's the whole point of the runtime check. if we do move this line one down, it will actually be a redundant assertion only intended to communicate with the compiler. a human reader already has enough context to know that (if you can wrap your head around the difference between `r` and `arr`; sorry for the confusing names.).

The compile error you're introducing seems worth this awkward workaround though. I think it would just be nice to have a comment on these redundant asserts explaining that they are working around a compile error.

Reply:
Good point - I'll move those checks down a line and add explanatory comments so readers are less confused by the assertions.

I wonder if we get ranged integers (https://github.com/ziglang/zig/issues/3806) if there would be any hope they would enable the compiler to statically know `arr.len <= r.len` after the runtime check.